### PR TITLE
Add soil temperature and moisture IAU

### LINKED ISSUE #2414: Add Land IAU Capability to NoahMP Driver in CCPP
## Description
The atmospheric DA applies IAU increments on temperature (among other variables), while the land surface increments are applied once at the start of the DA (cycle?) window. This could lead to the land surface (soil and snow) temperature being out of sync with near surface atmospheric temperature.

## Solution
Add a Land IAU module to calculate increments for soil temperature. Later on, this can be extended to include snow temperature, and if necessary, the other soil and snow variables. The module’s functions will be called either from within the noahmpdrv ccpp scheme or right before the noahmpdrv run.

## Alternatives
The existing atmospheric IAU module (fv_iau_mod.F90) can be modified with few additions for land, but because fv_iau_mod.F90 is part of the dynamical core there is strong resistance by repo owners. Having a separate land IAU module as part of ccpp’s Land_Model components (also closer to the schemes that use it) is a better option.

## Related to
- https://github.com/NOAA-EMC/fv3atm/issues/867
- https://github.com/ufs-community/ccpp-physics/issues/190
- https://github.com/NOAA-EMC/noahmp/pull/12

### PR DESCRIPTION
**Pull request by tsga**
## Commit Queue Requirements:

- [x] Fill out all sections of this template.
- [x] All sub component pull requests have been reviewed by their code managers.
- [x] Run the full Intel+GNU RT suite (compared to current baselines) on either Hera/Derecho/Hercules
- [x] Commit 'test_changes.list' from previous step
---
## Description:
Add Incremental Analysis Update (IAU) to soil temperature and moisture in the NoahMP land model of the UWM's CCPP.

### Commit Message:

```
* UFSWM 
  * FV3
    * ccpp-physics - Add Incremental Analysis Update (IAU) capability to update soil temperature and moisture for CCPP's NoahMP land model
```

### Priority:

* Normal

### UFSWM:

* Closes #2414 

### Sub component Pull Requests:

* FV3: NOAA-EMC/fv3atm/pull/866
  * ccpp-physics: ufs-community/ccpp-physics/pull/222
* noahmp: https://github.com/NOAA-EMC/noahmp/pull/12

### UFSWM Blocking Dependencies:

* None

---
### Regression Test Changes (Please commit test_changes.list):

* PR Adds New Tests/Baselines.
Additional land-IAU specific test to be added: control_c48_intel_land_iau (based on the "control_c48_intel" test, with the only difference being the turning on of land IAU and soil temperature increment files inside INPUT directory). The test inputs have been staged in hera platform. The baseline outputs created for the control_c48_lnd_iau_intel regression tests is at:
/scratch1/NCEPDEV/stmp4/Tseganeh.Gichamo/REGRESSION_TEST_LAND_IAU/control_c48_lnd_iau_intel

### Input data Changes:

* New input data.
soil temperature increment files inside INPUT directory

### Library Changes/Upgrades:

* No Updates
  
---

## Testing Log:
- RDHPCS
  - [x] Hera
  - [x] Orion
  - [x] Hercules
  - [x] Jet
  - [x] Gaea
  - [ ] Derecho
- WCOSS2
  - [x] Dogwood/Cactus
  - [ ] Acorn
- [ ] CI
- [ ] opnReqTest (complete task if unnecessary)

### CODE REVIEW

**Comment by tsga** [comment]
Tagging relevant people here. Please add (tag) anyone you think should be aware.
@CoryMartin-NOAA, @barlage, @ClaraDraper-NOAA, @CatherineThomas-NOAA, @junwang-noaa

**Comment by grantfirl** [comment]
@tsga There should be an associated PR for the NoahMP changes in https://github.com/NOAA-EMC/noahmp, right? This component repo needs to have its internal NoahMP code updated to correspond to the changes in https://github.com/ufs-community/ccpp-physics/pull/222. Please let me know if you need help with this or if I am mistaken.

**Comment by grantfirl** [comment]
@tsga Since there is a new namelist to read in (according to https://github.com/ufs-community/ccpp-physics/pull/222/files#diff-13b994b587cb8a841246d6db7df55ef9ff6fc1269777b9fe1b2e704ce9da629bR138), will an example of this be added to this repository for testing purposes?

**Comment by tsga** [comment]
> @tsga Since there is a new namelist to read in (according to https://github.com/ufs-community/ccpp-physics/pull/222/files#diff-13b994b587cb8a841246d6db7df55ef9ff6fc1269777b9fe1b2e704ce9da629bR138), will an example of this be added to this repository for testing purposes?

@grantfirl yes there is an example namelist with the new Land IAU specific test at:
/scratch1/NCEPDEV/da/Tseganeh.Gichamo/ccpp-physics/test/control_c48_intel_land_iau/input.nml

Should I also add it as part of this repo (as one of the files)?

**Comment by grantfirl** [comment]
> > @tsga Since there is a new namelist to read in (according to https://github.com/ufs-community/ccpp-physics/pull/222/files#diff-13b994b587cb8a841246d6db7df55ef9ff6fc1269777b9fe1b2e704ce9da629bR138), will an example of this be added to this repository for testing purposes?
> 
> @grantfirl yes there is an example namelist with the new Land IAU specific test at: /scratch1/NCEPDEV/da/Tseganeh.Gichamo/ccpp-physics/test/control_c48_intel_land_iau/input.nml
> 
> Should I also add it as part of this repo (as one of the files)?

@tsga @jkbk2004 As of now, if we merged your work, there would be no tests that are utilizing it, so it could be broken with subsequent code changes. You mention in this PR description that a new RT should be added, but the files haven't been modified to add it yet. You'd need to edit tests/rt.conf, commit the new namelist and refer to it in the new test file to be added in tests/tests, and have the new data staged on the testing platforms in order for the test to be set up to be exercised.

**Comment by grantfirl** [comment]
@tsga Please see https://github.com/tsga/ufs-weather-model/pull/1,
https://github.com/tsga/fv3atm/pull/1, and https://github.com/tsga/ccpp-physics/pull/1

With these changes, I can use the new RT via: `./rt.sh -e -k -n "control_c48_lnd_iau intel" > rt_single_control_c48_lnd_iau.log 2>&1 &`

It compiles successfully, although the run fails because the data is not staged. When the data is staged, we can complete testing.

**Comment by tsga** [comment]
@grantfirl Thank you! I added the increment files to the INPUT dir of the test you set up and it runs successfully (I am working on the other issues of gracefully exiting when there are no files, etc, you mentioned earlier). 
In the mean time, can you copy these files (sfc_inc*.nc) /scratch2/NCEPDEV/land/data/DA/soil_moisture/LandIAU/INC/C48/INC09/* 
to /scratch2/NAGAPE/epic/UFS-WM_RT/NEMSfv3gfs/input-data-20240501/FV3_input_data/INPUT_L127_gfsv17 (I don't have permission). These six files are the only ones needed to make the control_c48_lnd_iau intel test that you set up complete.

**Comment by grantfirl** [comment]
> @grantfirl Thank you! I added the increment files to the INPUT dir of the test you set up and it runs successfully (I am working on the other issues of gracefully exiting when there are no files, etc, you mentioned earlier). In the mean time, can you copy these files (sfc_inc*.nc) /scratch2/NCEPDEV/land/data/DA/soil_moisture/LandIAU/INC/C48/INC09/* to /scratch2/NAGAPE/epic/UFS-WM_RT/NEMSfv3gfs/input-data-20240501/FV3_input_data/INPUT_L127_gfsv17 (I don't have permission). These six files are the only ones needed to make the control_c48_lnd_iau intel test that you set up complete.

@tsga I don't have permissions in that directory either. @jkbk2004 and/or other members of the UFS code management team can do this for us.

**Comment by jkbk2004** [comment]
> > @grantfirl Thank you! I added the increment files to the INPUT dir of the test you set up and it runs successfully (I am working on the other issues of gracefully exiting when there are no files, etc, you mentioned earlier). In the mean time, can you copy these files (sfc_inc*.nc) /scratch2/NCEPDEV/land/data/DA/soil_moisture/LandIAU/INC/C48/INC09/* to /scratch2/NAGAPE/epic/UFS-WM_RT/NEMSfv3gfs/input-data-20240501/FV3_input_data/INPUT_L127_gfsv17 (I don't have permission). These six files are the only ones needed to make the control_c48_lnd_iau intel test that you set up complete.
> 
> @tsga I don't have permissions in that directory either. @jkbk2004 and/or other members of the UFS code management team can do this for us.

@tsga @grantfirl I will follow up today. will keep you posted.

**Comment by jkbk2004** [comment]
@tsga /scratch2/NAGAPE/epic/UFS-WM_RT/NEMSfv3gfs/input-data-20240501/FV3_input_data/INPUT_L127_gfsv17 is for C96 resolution. Do you want me to copy over /scratch2/NCEPDEV/land/data/DA/soil_moisture/LandIAU/INC/C96/sfc_inc.tile*.nc ?

**Comment by jkbk2004** [comment]
@tsga Or if you want to stay with with C48, we have C48 test case at /scratch2/NAGAPE/epic/UFS-WM_RT/NEMSfv3gfs/input-data-20240501/FV3_input_data48/INPUT_L127_gfsv17. We can copy over your C48 sfc increment files there. Maybe both C48 and C96. Let me know.

**Comment by tsga** [comment]
> @tsga Or if you want to stay with with C48, we have C48 test case at /scratch2/NAGAPE/epic/UFS-WM_RT/NEMSfv3gfs/input-data-20240501/FV3_input_data48/INPUT_L127_gfsv17. We can copy over your C48 sfc increment files there. Maybe both C48 and C96. Let me know.

@jkbk2004 yes, let's stay with C48. And disregard the earlier directory. The C48 increments are at: /scratch2/NCEPDEV/land/data/DA/soil_moisture/LandIAU/INC/C48/INC3

**Comment by jkbk2004** [comment]
@tsga /scratch2/NAGAPE/epic/UFS-WM_RT/NEMSfv3gfs/input-data-20240501/FV3_input_data48/INPUT_L127_gfsv17/sfc_inc.tile*.nc is available. Once test is ok on hera, then we can rsync up on other machines.

**Comment by tsga** [comment]
> @tsga There should be an associated PR for the NoahMP changes in https://github.com/NOAA-EMC/noahmp, right? This component repo needs to have its internal NoahMP code updated to correspond to the changes in [ufs-community/ccpp-physics#222](https://github.com/ufs-community/ccpp-physics/pull/222). Please let me know if you need help with this or if I am mistaken.

@barlage I would like to hear your opinion here. My impulse is to finalize the current pull request for ccpp-physics/fv3/ufs-wm and then start work for the component model separately. 

@grantfirl I am assuming there is no repo dependence here (like that exists between ccpp-physics->fv3->ufswm)? Whatever will be merged for ccpp-physics/noahmp will be ported to the NOAA-EMC/noahmp.

**Comment by tsga** [comment]
> @tsga /scratch2/NAGAPE/epic/UFS-WM_RT/NEMSfv3gfs/input-data-20240501/FV3_input_data48/INPUT_L127_gfsv17/sfc_inc.tile*.nc is available. Once test is ok on hera, then we can rsync up on other machines.

@jkbk2004 I can now confirm that the land iau tests work on hera. The new baseline created for the "control_c48_lnd_iau_intel" test is at:
/scratch1/NCEPDEV/stmp4/Tseganeh.Gichamo/FV3_RT/REGRESSION_TEST/control_c48_lnd_iau_intel

**Comment by CoryMartin-NOAA** [comment]
> > @tsga There should be an associated PR for the NoahMP changes in https://github.com/NOAA-EMC/noahmp, right? This component repo needs to have its internal NoahMP code updated to correspond to the changes in [ufs-community/ccpp-physics#222](https://github.com/ufs-community/ccpp-physics/pull/222). Please let me know if you need help with this or if I am mistaken.
> 
> @barlage I would like to hear your opinion here. My impulse is to finalize the current pull request for ccpp-physics/fv3/ufs-wm and then start work for the component model separately.
> 
> @grantfirl I am assuming there is no repo dependence here (like that exists between ccpp-physics->fv3->ufswm)? Whatever will be merged for ccpp-physics/noahmp will be ported to the NOAA-EMC/noahmp.

Would the component model require changes to build with the changes to CCPP, or just require changes to use the land IAU? If the former, then the PRs should probably be concurrent, but if the latter, then I support @tsga 's impulse to finish this first, and then move on to the component model.

**Comment by barlage** [comment]
> > > @tsga There should be an associated PR for the NoahMP changes in https://github.com/NOAA-EMC/noahmp, right? This component repo needs to have its internal NoahMP code updated to correspond to the changes in [ufs-community/ccpp-physics#222](https://github.com/ufs-community/ccpp-physics/pull/222). Please let me know if you need help with this or if I am mistaken.
> > 
> > 
> > @barlage I would like to hear your opinion here. My impulse is to finalize the current pull request for ccpp-physics/fv3/ufs-wm and then start work for the component model separately.
> > @grantfirl I am assuming there is no repo dependence here (like that exists between ccpp-physics->fv3->ufswm)? Whatever will be merged for ccpp-physics/noahmp will be ported to the NOAA-EMC/noahmp.
> 
> Would the component model require changes to build with the changes to CCPP, or just require changes to use the land IAU? If the former, then the PRs should probably be concurrent, but if the latter, then I support @tsga 's impulse to finish this first, and then move on to the component model.

I believe there will be an issue here due to change in the noahmpdrv_init subroutine. I set up a meeting with @uturuncoglu  to discuss.

**Review by ClaraDraper-NOAA — APPROVED** [review]
looks good to me, thanks @tsga

**Comment by jkbk2004** [comment]
@tsga can you sync up branch? We can start working on this pr.

**Comment by tsga** [comment]
@jkbk2004 Thanks for fixing the git submodules and syncing up.
I have re-run the tests with the latest changes. Everything looks good and the sub-repos are cleaned up.
The baseline created for the control_c48_lnd_iau_intel regression tests is at:
/scratch1/NCEPDEV/stmp4/Tseganeh.Gichamo/REGRESSION_TEST_LAND_IAU/control_c48_lnd_iau_intel

File: NOAHMP-interface/CMakeLists.txt [hunk 1]
```diff
@@ -13,7 +13,8 @@ list(APPEND _noahmp_cap_files noahmp/drivers/nuopc/lnd_comp_kind.F90
 
 # CCPP interface
 list(APPEND _noahmp_ccpp_files noahmp/drivers/ccpp/noahmpdrv.F90
-                               noahmp/drivers/ccpp/sfc_diff.f
+	                       noahmp/drivers/ccpp/lnd_iau_mod.F90
+	                       noahmp/drivers/ccpp/sfc_diff.f
```
Comment (BrianCurtis-NOAA) [inline NOAHMP-interface/CMakeLists.txt]: Spacing here is inconsistent

**Review by BrianCurtis-NOAA — CHANGES_REQUESTED** [review]
I'm confused here. There is a new test that creates new baselines, yet there's no entries in test_changes.list. This should have been taken care of before testing started.

**Comment by jkbk2004** [comment]
> I'm confused here. There is a new test that creates new baselines, yet there's no entries in test_changes.list. This should have been taken care of before testing started.

@BrianCurtis-NOAA this pr adds new test case control_c48_lnd_iau intel. I pushed test_changes.list

**Comment by BrianCurtis-NOAA** [comment]
> > I'm confused here. There is a new test that creates new baselines, yet there's no entries in test_changes.list. This should have been taken care of before testing started.
> 
> @BrianCurtis-NOAA this pr adds new test case control_c48_lnd_iau intel. I pushed test_changes.list

@jkbk2004 It looks like there was a full suite tested on hera: https://github.com/ufs-community/ufs-weather-model/pull/2415/commits/d93cf7a8e503bf9cb21914053ae76f715ffbe82f but the new test never made it into that log, which might be why the test_changes.list was empty. Make sure every commit is reviewed in some capacity to help ensure everything is good to go prior to starting testing.

File: NOAHMP-interface/CMakeLists.txt [hunk 2]
```diff
@@ -13,7 +13,8 @@ list(APPEND _noahmp_cap_files noahmp/drivers/nuopc/lnd_comp_kind.F90
 
 # CCPP interface
 list(APPEND _noahmp_ccpp_files noahmp/drivers/ccpp/noahmpdrv.F90
-                               noahmp/drivers/ccpp/sfc_diff.f
+	                             noahmp/drivers/ccpp/lnd_iau_mod.F90
+	                             noahmp/drivers/ccpp/sfc_diff.f
                                noahmp/drivers/ccpp/machine.F
```
Comment (BrianCurtis-NOAA) [inline NOAHMP-interface/CMakeLists.txt]: It might be a spaces vs tabs thing, but it's still not lined up here.

**Review by BrianCurtis-NOAA — CHANGES_REQUESTED** [review]

**Comment by BrianCurtis-NOAA** [comment]
Skip Acorn

**Review by jkbk2004 — APPROVED** [review]

**Review by FernandoAndrade-NOAA — APPROVED** [review]

File: NOAHMP-interface/CMakeLists.txt (hunk 2 above)
Comment (FernandoAndrade-NOAA) [reply NOAHMP-interface/CMakeLists.txt]: @BrianCurtis-NOAA it looks like you may need to approve this/mark as resolved before merge. The changes should already be in.

**Review by BrianCurtis-NOAA — APPROVED** [review]

**Comment by tsga** [comment]
@jkbk2004 and others, thank you so much for your help and guidance through this PR. Cheers.

**Pull request merged**